cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/CapacityGoal.java

```diff
@@ -153,10 +153,23 @@ protected void initGoalState(ClusterModel clusterModel, OptimizationOptions opti
     double allowedCapacity = capacity * _balancingConstraint.capacityThreshold(resource());
 
     if (allowedCapacity < existingUtilization) {
+      Set<Integer> brokersAllowedReplicaMove = GoalUtils.aliveBrokersNotExcludedForReplicaMove(clusterModel, optimizationOptions);
+      if (brokersAllowedReplicaMove.isEmpty()) {
+        // Handle the case when all alive brokers are excluded from replica moves.
+        throw new OptimizationFailureException(String.format("[%s] All alive brokers are excluded from replica moves.", name()),
+                                               String.format("Add at least %d brokers.", clusterModel.maxReplicationFactor()));
+      }
+
+      // Identify a typical broker capacity to be used in recommendations in case the cluster is under-provisioned.
+      int typicalBrokerId = brokersAllowedReplicaMove.iterator().next();
+      double typicalCapacity = clusterModel.broker(typicalBrokerId).capacityFor(resource());
+      double missingCapacity = existingUtilization - allowedCapacity;
+      int numBrokersToAdd = (int) (missingCapacity / (typicalCapacity * _balancingConstraint.capacityThreshold(resource())));
+
       throw new OptimizationFailureException(
```

Comment:
Exceptions are very, very low performance and should only be used for exceptional circumstances.   It seems like having a recommendation here is a normal outcome.  A normal, large scale Kafka deployment will probably have some clusters that are not optimally sized.  If that's the case then may the return type needs to be changed.  In c++ I might used a variant type that can either be boolean or String (the recommendation).  A quick search for Java variant doesn't turn up anything interesting. 
```java
class Optimization {
  private Boolean opt;
  private String recommendation;

 // boolean constructor
 // String constructor
 void complete(Function<Boolean> onOptimization, Function<String> onRecommendation) {
   if (opt != null) {
     onOptimization(opt);
   } else if (recommendation != null) {
     onRecommendation(recommendation);
   } else {
    //constructor should prevent this from happening
  }
 } 
}
```

Reply:
Thanks for the detailed explanation and the example case! I agree that exceptions are expensive and should be avoided when possible -- so that is a great general advice.

That being said, an `OptimizationFailureException` is expected to be a rare event -- i.e. unless it is fixed, Cruise Control cannot function. For example, it cannot manage the cluster to perform operations including rebalance, broker removal or addition, fixing offline replicas, and changing replication factor. 

      // An OptimizationFailureException indicates (1) a hard goal violation that cannot be fixed typically due to
      // lack of physical hardware (e.g. insufficient number of racks to satisfy rack awareness, insufficient number
      // of brokers to satisfy Replica Capacity Goal, or insufficient number of resources to satisfy resource
      // capacity goals), or (2) a failure to move offline replicas away from dead brokers/disks.

I see that a part of this confusion might be due to what CC considers as "under-provisioned". The current iteration of CC is quite conservative on what it considers as under-provisioned -- i.e. CC considers only the clusters that lead to an `OptimizationFailureException` for a goal as under-provisioned. So just a non-optimally sized cluster is unlikely to be reported as under-provisioned.

Reply:
Hmm.  I'm not really convinced and I think this will lead to poor performance.  I guess we will find out.

Reply:
Does the exception throw cause termination of the goal?  Will I miss other optimizations that might be generated if there was a normal return?

Reply:
> Does the exception throw cause termination of the goal?

Yes, the optimization for the goal will terminate with this exception message.

> Will I miss other optimizations that might be generated if there was a normal return?

There are 2 parts to the response:
1. Requests via REST API, which require an optimization, will (as before) receive the exception message with the optimization failure details for the first goal with `OptimizationFailureException` having the highest priority.
2. Anomaly Detector Substate > Goal violation will show the information regarding provision recommendations for each goal with optimization failure exception (if any).

> Hmm. I'm not really convinced and I think this will lead to poor performance. I guess we will find out.

This patch does not introduce or increase the frequency of these exceptions -- these exceptions were already in use before. So it is not clear why we would find out a performance difference due to this patch.